streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,639 @@
+            if self.validate_hash:
+                data = open(raw_filename, 'rb').read()
+                assert get_hash(self.validate_hash, data) == raw_info.hashes[self.validate_hash]
+
+    def _download_shard(self, shard_id: int) -> None:
+        """Download the given shard.
+
+        Args:
+            shard_id (int): Shard ID.
+        """
+        reader = self.shards[shard_id]
+        for raw_info, zip_info in reader.file_pairs:
+            self._download_shard_part(shard_id, raw_info, zip_info, reader.compression)
+
+    def _download_or_await_shard(self, lock: FileLock, shard_states: NDArray[np.uint8],
+                                 shard_id: int) -> None:
+        """Either download the given shard or wait on its download.
+
+        Args:
+            lock (FileLock): The lock protecting ``shard_states``.
+            shard_states (NDArray[np.uint8]): The download status of each shard, as an array in
+                shared memory.
+            shard_id (int): Shard ID.
+        """
+        # First, the fast path: check the shared memory shard state without taking the lock. The
+        # shard states only ever go up, so if we're at the downloaded state, it's downloaded.
+        state = shard_states[shard_id]
+        if state == _ShardState.DOWNLOADED:
+            return
+
+        # Shard is not necessarily downloaded, so check and update state with the lock.
+        lock.acquire()
+        state = shard_states[shard_id]
+        if state == _ShardState.UNKNOWN:
+            shard_states[shard_id] = _ShardState.DOWNLOADING
+            lock.release()
+            self._download_shard(shard_id)
+            with lock:
+                shard_states[shard_id] = _ShardState.DOWNLOADED
+        elif state == _ShardState.DOWNLOADING:
+            lock.release()
+            while shard_states[shard_id] != _ShardState.DOWNLOADED:
+                sleep(0.07)
+        elif state == _ShardState.DOWNLOADED:
+            lock.release()
+        else:
+            raise RuntimeError('Unknown shard state')
+
+    def _download_thread(self, epoch: int, sample_ids: NDArray[np.int64]) -> None:
+        """Download the relevant shards in the background while we are being iterated.
+
+        Args:
+            epoch (int): Which epoch. On noticing that a new epoch has started, we exit this thread
+                because a new one will soon be running, with different sample IDs.
+            sample_ids (NDArray[np.int64]): The samples to download the shards of.
+        """
+        # Create or attach shard_states array.
+        name = f'{self._prefix}_shard_states'
+        size = len(self.shard_sizes) * np.uint8(0).nbytes
+        try:
+            shm = SharedMemory(name, True, size)
+        except:
+            shm = SharedMemory(name)
+        shard_states = np.ndarray(len(self.shard_sizes), buffer=shm.buf, dtype=np.uint8)
+
+        filename = os.path.join('/tmp', 'mds', self._prefix, '_shard_states_filelock')
```

Comment:
Few things:
1. use `os.sep` instead of `/`.
2. why `mds` ? since we support multiple writers of different type. Can we get `mds` name during runtime ?

Reply:
Fixed